dm/tests/shardddl_optimistic/run.sh

```diff
@@ -0,0 +1,476 @@
+	run_sql_source2 "insert into ${shardddl1}.${tb2} values(12,'1212','121212');"
+
+	run_sql_source2 "alter table ${shardddl1}.${tb2} drop column b;"
+	random_restart
+
+	# source1.tb1(a,c); source1.tb2(a,b); source2.tb1(a,c); source2.tb2(a,c)
+	run_sql_source1 "insert into ${shardddl1}.${tb1} values(13,'131313');"
+	run_sql_source1 "insert into ${shardddl1}.${tb2} values(14,'1414');"
+	run_sql_source2 "insert into ${shardddl1}.${tb1} values(15,'151515');"
+	run_sql_source2 "insert into ${shardddl1}.${tb2} values(16,'161616');"
+
+	run_sql_source1 "alter table ${shardddl1}.${tb2} drop column b;"
+	random_restart
+
+	# source1.tb1(a,c); source1.tb2(a); source2.tb1(a,c); source2.tb2(a,c)
+	run_sql_source1 "insert into ${shardddl1}.${tb1} values(17,'171717');"
+	run_sql_source1 "insert into ${shardddl1}.${tb2} values(18);"
+	run_sql_source2 "insert into ${shardddl1}.${tb1} values(19,'191919');"
+	run_sql_source2 "insert into ${shardddl1}.${tb2} values(20,'202020');"
+
+	run_sql_source1 "alter table ${shardddl1}.${tb2} add column c text;"
+	random_restart
+
+	# source1.tb1(a,c); source1.tb2(a,c); source2.tb1(a,c); source2.tb2(a,c)
+	run_sql_source1 "insert into ${shardddl1}.${tb1} values(21,'212121');"
+	run_sql_source1 "insert into ${shardddl1}.${tb2} values(22,'222222');"
+	run_sql_source2 "insert into ${shardddl1}.${tb1} values(23,'232323');"
+	run_sql_source2 "insert into ${shardddl1}.${tb2} values(24,'242424');"
+
+	run_sql_tidb_with_retry "select count(1) from ${shardddl}.${tb}" "count(1): 24"
+	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+}
+
+function DM_RESTART_TASK_MASTER_WORKER() {
+	# create table with different schema, init data, and create table in downstream manually
+	run_case RESTART_TASK_MASTER_WORKER "double-source-optimistic" \
+		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key);\"; \
+   	run_sql_source1 \"create table ${shardddl1}.${tb2} (a int primary key, b varchar(10));\"; \
+   	run_sql_source2 \"create table ${shardddl1}.${tb1} (a int primary key, c text);\"; \
+   	run_sql_source2 \"create table ${shardddl1}.${tb2} (a int primary key, b varchar(10), c text);\"; \
+   	run_sql_source1 \"insert into ${shardddl1}.${tb1} values(1);\"; \
+   	run_sql_source1 \"insert into ${shardddl1}.${tb2} values(2,'22');\"; \
+   	run_sql_source2 \"insert into ${shardddl1}.${tb1} values(3,'333');\"; \
+   	run_sql_source2 \"insert into ${shardddl1}.${tb2} values(4,'44','444');\"; \
+   	run_sql_tidb \"create database if not exists ${shardddl};\"; \
+   	run_sql_tidb \"create table ${shardddl}.${tb} (a int primary key, b varchar(10), c text);\"" \
+		"clean_table" "optimistic"
+}
+
+function random_restart() {
+	mod=$(($RANDOM % 4))
+	if [[ "$mod" == "0" ]]; then
+		restart_master
+	elif [[ "$mod" == "1" ]]; then
+		restart_worker1
+	elif [[ "$mod" == "2" ]]; then
+		restart_worker2
+	else
+		restart_task $cur/conf/double-source-optimistic.yaml
+	fi
+}
+
```

Comment:
Actually, I think it's better to test both four cases for integration tests. It will be easier for us to trace which PR introduces the bug.

Reply:
I want to test different combinations of them. e.g.(restart master and then restart worker), but too many combinations, so use randome. I add some log for trace in [41cf696](https://github.com/pingcap/tiflow/pull/3903/commits/41cf696a23d1041a75b4cfc70d05f1a4e8cd0c41)

Reply:
I see